src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -563,65 +573,135 @@ class segment_meta_cstore::impl
       segment_meta_cstore::impl,
       serde::version<0>,
       serde::compat_version<0>> {
+    // TODO tunable?
+    constexpr static auto max_buffer_entries = 1024u;
+
 public:
+    // TODO bypass buffer?
     void append(const segment_meta& meta) { _col.append(meta); }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> begin() const {
+        // TODO deal with buffer
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.begin());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> end() const {
+        // TODO deal with buffer
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.end());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     find(model::offset o) const {
+        // TODO deal with buffer
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.find(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     lower_bound(model::offset o) const {
+        // TODO deal with buffer
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.lower_bound(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     upper_bound(model::offset o) const {
+        // TODO deal with buffer
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.upper_bound(o()));
     }
 
     std::optional<segment_meta> last_segment() const {
+        if (!write_buffer.empty()) {
+            auto last_committed_offset = _col.last_committed_offset().value_or(
+              model::offset::min());
+            auto buffer_last_base_offset = write_buffer.crbegin()->first;
+            // CHECK: base_offset is never equal to committed_offset
+            //
+            if (likely(buffer_last_base_offset > last_committed_offset)) {
+                return write_buffer.crbegin()->second;
+            }
+            // fallthrough
+        }
         return _col.last_segment();
     }
 
-    void insert(const segment_meta& m) { _col.append(m); }
+    void insert(const segment_meta& m) {
+        // TODO optimize to bypass this write?
+        write_buffer[m.base_offset] = m;
+        if (write_buffer.size() > max_buffer_entries) {
+            flush_write_buffer();
+        }
+    }
 
-    auto inflated_actual_size() const { return _col.inflated_actual_size(); }
+    auto inflated_actual_size() const {
+        // TODO how to deal with write_buffer?
+        return _col.inflated_actual_size();
+    }
 
-    size_t size() const { return _col.size(); }
+    size_t size() const {
+        // TODO how to deal with write_buffer?
```

Comment:
I'm still on the fence about this. Since the result could be used for at_index, it needs to be exact. On the other hand, at_index does not look like a core operation that needs to be there, so it could be moved to a private context, and size could be renamed to "approximate_size"